src/main/java/roboy/context/Context.java

```diff
@@ -1,135 +1,197 @@
+    public final ActiveInterlocutorUpdater ACTIVE_INTERLOCUTOR_UPDATER;
+
+    /* EXTERNAL UPDATERS DEFINED HERE */
+    private volatile boolean rosInitialized = false;
+    private AudioDirectionUpdater AUDIO_ANGLES_UPDATER;
+    private ROSTestUpdater ROS_TEST_UPDATER;
+
+    /* OBSERVERS DEFINED HERE */
+    private final FaceCoordinatesObserver FACE_COORDINATES_OBSERVER;
+
+
+    // By calling getInstance, Context initializes its updaters and observers.
+    private static Context context = Context.getInstance();
+
+    /**
+     * Builds the class to instance maps.
+     */
     private Context() {
-        FaceCoordinates faceCoordinates = new FaceCoordinates();
-        DialogTopics dialogTopics = new DialogTopics();
-
-        values = new ImmutableClassToInstanceMap.Builder<AbstractValue>()
-                // Collect all Values centrally right here.
-                .put(FaceCoordinates.class, faceCoordinates)
-                .build();
-
-        valueHistories = new ImmutableClassToInstanceMap.Builder<AbstractValueHistory>()
-                // Collect all ValueHistories centrally right here.
-                .put(DialogTopics.class, dialogTopics)
-                .build();
-
-        // Initialize and store the external updaters.
-        externalUpdaters.add(new FaceCoordinatesUpdater(faceCoordinates, 1));
-        // Initialize direct updaters with their targets.
-        internalUpdaters.put(DialogTopicsUpdater.class, new DialogTopicsUpdater(dialogTopics));
+        /* INTERNAL UPDATERS INITIALIZED HERE */
+        DIALOG_TOPICS_UPDATER = new DialogTopicsUpdater(DIALOG_TOPICS.valueHistory);
+        DIALOG_INTENTS_UPDATER = new DialogIntentsUpdater(DIALOG_INTENTS.valueHistory);
+        ACTIVE_INTERLOCUTOR_UPDATER = new ActiveInterlocutorUpdater(ACTIVE_INTERLOCUTOR.value);
+
+        /* OBSERVERS INITIALIZED AND ADDED HERE */
+        FACE_COORDINATES_OBSERVER = new FaceCoordinatesObserver();
+//        FACE_COORDINATES.value.addObserver(FACE_COORDINATES_OBSERVER);
     }
 
+    /**
+     * Starts up the external updaters (which need a ROS main node).
+     * @param ros
+     */
+    public void initializeROS(RosMainNode ros) {
+        Context ctx = Context.getInstance();
+        synchronized (initializationLock) {
+            // Initialize only if not already done.
+            if(!rosInitialized) {
+
+                /* EXTERNAL UPDATERS INITIALIZED HERE */
+//                AUDIO_ANGLES_UPDATER = new AudioDirectionUpdater(AUDIO_ANGLES.valueHistory, ros);
+//                ROS_TEST_UPDATER = new ROSTestUpdater(ROS_TEST.valueHistory, ros);
+                // TODO Add a FACE_COORDINATE_UPDATER.
+                // Edit the data type and integration tests, once the real data type is used.
+
+                rosInitialized = true;
+            }
+        }
+    }
+
+    /**
+     * The access point to Context, including thread-safe Singleton initialization.
+     */
     public static Context getInstance() {
```

Comment:
Alona was right about singletons. In the future, we might need multiple independent Contexts, for example if we like to implement a chatbot that has multiple conversations at once. This is obviously not possible with a singleton context. 

However, don't worry about this. It is out of scope for this semester. New students in the next one also need some tasks to do ;)

Reply:
I would first analyze if multiple Contexts are really-really necessary. Originally, the data in Context was meant to be globally accessible, which is why I made sure that there would only be one instance of Context floating around during runtime, not multiple ones containing partial data ;)

I don't have a good estimation about how much of the Context data would be conversation-specific and require separate storage. If it is an exception rather than the rule, we could also experiment with creating unique IDs for each conversation and storing conversation-specific data in a map-type Value.

Reply:
Context data is always conversation specific

Reply:
Perception data would be global.